### DISCUSSION TITLE
How do you use a generic type in a class-based dependency?

### DISCUSSION THREAD

**Discussion by usrbinsam** [fastapi/fastapi#8782] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from dataclasses import dataclass
from typing import TypeVar, Generic

from fastapi import FastAPI, Depends
from sqlalchemy import create_engine, Column, Integer, String
from sqlalchemy.orm import declarative_base, sessionmaker, Session
from starlette.exceptions import HTTPException

# Models
M = TypeVar("M")
S = TypeVar("S")

engine = create_engine("sqlite:///:memory:")
Base = declarative_base()
SessionLocal = sessionmaker(bind=engine)


class MyModel(Base):
    __tablename__ = "my_model"
    id = Column(Integer, primary_key=True)
    name = Column(String(32), nullable=False)


class MySchema:
    id: int
    name: str


@dataclass
class Result(Generic[M, S]):
    entity: M
    schema: S


Base.metadata.create_all(bind=engine)


# Dependencies
def get_db() -> Session:
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


class EntityGetter(Generic[S]):

    def __init__(self, model_type: Base):
        self.model_type = model_type

    def __call__(self, schema: S, db: Session = Depends(get_db)):
        model = db.get(self.model_type, schema.id)
        if model is None:
            raise HTTPException(status_code=404)
        yield Result(entity=model, schema=schema)
        db.commit()


# Routes
app = FastAPI()


@app.on_event("startup")
def create():
    m = MyModel(name="foo")
    db = SessionLocal()
    db.add(m)
    db.commit()


GetModel = EntityGetter[MySchema](MyModel)


@app.put("/model")
def update_model(result: Result[MyModel, MySchema] = Depends(GetModel)):
    result.entity.name = result.schema.name
    return result.entity


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, port=9000)
```

### Description

Hi, I'm trying to use Generics in class-based dependencies. I made a trivial example to show how I'm trying to use it, but FastAPI does not seem to recognize the Generic type `S`.  Perhaps this is beyond the current capabilities of FastAPI, or I'm using Generics improperly (kinda new to Python's typing system). 

Any pointers in the correct direction here would be greatly appreciated. The goal with this Dependency was to use it to use the `id` attribute of the given schema to lookup the corresponding ORM model and return both in a Result container so they can be manipulated in the route handler.

For now I will probably just have my dependency grab `request.json()["id"]` as a workaround instead of using a Generic. 

### Operating System

Windows

### FastAPI Version

0.68.0

### Python Version

Python 3.9.6

### Additional Context

![image](https://user-images.githubusercontent.com/342758/128782902-01dd5736-3f9e-4d8c-9ee6-341ae4ef04e8.png)

**Comment by usrbinsam** [comment]
Decided I like the [service](https://www.patrick-muehlbauer.com/articles/fastapi-with-sqlalchemy) pattern better for this use case, but will leave the question open for some insight on generics within a dependency.

**Reply by flo-at** [reply]
Are you still using this pattern? Looking at the repository from the blog post you referenced, I'm trying to find a solution to reduce the duplication in [this directory](https://github.com/treebee/orders-api/tree/main/src/orders_api/routers) where basically all the files are the same except for the name of the resources.

**Comment by adriangb** [comment]
I don't think this can ever work:

```python
from inspect import signature

from typing import Generic, TypeVar

T = TypeVar("T")

class GenericThing(Generic[T]):

    def func(self, thing: T) -> None:
        ...

class ConstrainedThing(GenericThing[int]):
    ...

c = ConstrainedThing()
assert signature(c.func).parameters["thing"].annotation == T
```

Information on generic binding gets erased/dropped at runtime. Thus FastAPI can not determine the type of `schema` in `EntityGetter.__call__`.

**Comment by adriangb** [comment]
Well it looks like I've been proven wrong by @Fatal1ty

```python
from inspect import signature
from typing import Generic, TypeVar

from mashumaro.meta.helpers import resolve_type_vars  # pip install mashumaro

T = TypeVar("T")

class GenericThing(Generic[T]):

    def __call__(self, thing: T) -> None:
        ...

class ConstrainedThing(GenericThing[int]):
    ...

c = ConstrainedThing()
resolved = resolve_type_vars(ConstrainedThing)[GenericThing]
assert resolved[signature(c.__call__).parameters["thing"].annotation] == int
```
